File: app/src/main/java/com/maye/today/today/ui/activity/LoginActivity.java
package com.maye.today.today.ui.activity;

import android.app.Activity;
import android.os.Bundle;
import android.widget.TextView;

import com.jakewharton.rxbinding.view.RxView;
import com.maye.today.today.R;
import com.maye.today.today.login.LoginPresenter;
import com.maye.today.today.login.LoginPresenterImpl;
import com.maye.today.today.login.LoginView;
import com.rengwuxian.materialedittext.MaterialEditText;

import java.util.concurrent.TimeUnit;

import rx.functions.Action1;

/**
 * 登录Activity
 */
public class LoginActivity extends Activity implements LoginView {

    private LoginPresenter loginPresenter;
    private MaterialEditText met_username;
    private MaterialEditText met_password;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        initComponent();
    }

    /**
     * 初始化控件
     */
    private void initComponent() {
        met_username = (MaterialEditText) findViewById(R.id.met_username);
        met_password = (MaterialEditText) findViewById(R.id.met_password);

        TextView tv_login = (TextView) findViewById(R.id.tv_login);
        RxView.clicks(tv_login).
                throttleWithTimeout(300, TimeUnit.MILLISECONDS).
                subscribe(new Action1<Void>() {
                    @Override
                    public void call(Void aVoid) {
                        //获取用户输入
                        String username = met_username.getText().toString();
                        String password = met_password.getText().toString();

                        //执行登录
                        loginPresenter.loginCheck(username, password);


                    }
                });


        loginPresenter = new LoginPresenterImpl(this);
    }

    @Override
    public void usernameError(String error) {
        met_username.setError(error);
    }

    @Override
    public void passwordError(String error) {
        met_password.setError(error);
    }

    @Override
    public void showProgress() {

    }

    @Override
    public void hideProgress() {

    }
}

File: LoginModule/src/main/java/com/maye/login/ui/LoginActivity.kt
package com.maye.login.ui

import android.animation.ValueAnimator
import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.MotionEvent
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import com.afollestad.materialdialogs.MaterialDialog
import com.google.android.material.textfield.TextInputLayout
import com.maye.base.expand.safeClick
import com.maye.base.utils.Md5Util
import com.maye.login.R
import com.maye.login.utils.InputUtil
import kotlinx.android.synthetic.main.activity_login.*

/**
 * 登录Activity
 */
class LoginActivity : Activity(), com.maye.login.mvp.LoginView, View.OnTouchListener, View.OnClickListener {
    private var loginPresenter: com.maye.login.mvp.LoginPresenter? = null
    private var til_login_username: TextInputLayout? = null
    private var til_login_password: TextInputLayout? = null
    private var progressDialog: MaterialDialog? = null
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)
        initComponent()
    }

    override fun onResume() {
        super.onResume()
        til_login_username?.editText?.clearFocus()
        til_login_username?.error = null
        til_login_password?.editText?.clearFocus()
        til_login_password?.error = null
        setCustomerAnimation(iv_circle_inner, iv_circle_middle, iv_circle_outside)
    }

    /**
     * 初始化控件
     */
    private fun initComponent() {
        til_login_username = findViewById<View>(R.id.til_login_username) as TextInputLayout
        til_login_username?.editText?.setOnTouchListener(this)
        til_login_password = findViewById<View>(R.id.til_login_password) as TextInputLayout
        til_login_password?.editText?.setOnTouchListener(this)
        progressDialog = MaterialDialog.Builder(this).title("登录中").content("正在验证用户名及密码").progress(true, 0).build()
        val tv_login = findViewById<View>(R.id.tv_login) as TextView
        val tv_login_register = findViewById<View>(R.id.tv_login_register) as TextView
        tv_login_register.setOnClickListener(this)
        loginPresenter = com.maye.login.mvp.LoginPresenterImpl(this)

        //点击登录——防止多触
        tv_login.safeClick {
            //获取用户输入
            if (InputUtil.checkTextInputLayout(til_login_username) && InputUtil.checkTextInputLayout(til_login_password)) {
                val username = til_login_username?.editText?.text.toString()
                val password = til_login_password?.editText?.text.toString()

                //执行登录
                loginPresenter?.loginCheck(username, Md5Util.textToMd5(password))
            }
        }
    }

    /**
     * 设置自定义动画，使Sign周边环圈做位移动画
     */
    private fun setCustomerAnimation(inner: ImageView, middle: ImageView, outside: ImageView) {
        val valueAnimator = ValueAnimator.ofFloat(0f, 1.5f).setDuration(3000)
        valueAnimator.repeatMode = ValueAnimator.RESTART
        valueAnimator.repeatCount = ValueAnimator.INFINITE
        valueAnimator.addUpdateListener { animation ->
            val value1 = animation.animatedValue as Float
            val value2 = value1 + 0.2f
            val value3 = value1 + 0.4f
            val alpha1 = 2f * (1 - value1) + 1
            val alpha2 = 2f * (1 - value2) + 1
            val alpha3 = 2f * (1 - value3) + 1
            inner.scaleX = value1
            inner.scaleY = value1
            inner.alpha = alpha1

            middle.scaleX = value2
            middle.scaleY = value2
            middle.alpha = alpha2

            outside.scaleX = value3
            outside.scaleY = value3
            outside.alpha = alpha3
        }
        valueAnimator.start()
    }

    override fun inputError(index: String, error: String) {
        when (index) {
            "USERNAME" -> til_login_username?.error = error
            "PASSWORD" -> til_login_password?.error = error
        }
    }

    override fun showProgress(visible: Boolean) {
        if (visible) {
            progressDialog?.show()
        } else {
            progressDialog?.dismiss()
        }
    }

    override fun startActivity() {
        //TODO ARouter
//        startActivity(Intent(this, HomeActivity::class.java))
    }

    override fun onTouch(view: View, motionEvent: MotionEvent): Boolean {
        when (view.id) {
            R.id.et_login_username -> til_login_username?.error = null
            R.id.et_login_password -> til_login_password?.error = null
        }
        return false
    }

    override fun onClick(view: View) {
        when (view.id) {
            //TODO ARouter
//            R.id.tv_login_register -> startActivity(Intent(this, RegisterActivity::class.java))
        }
    }
}